### PR TITLE
Improve teamcity DSL code for bucket generation

### PR DESCRIPTION
**Pull request by wolfs**
After these changes, the functional tests from the buckets are only created in one place - no need to repeat the same code over and over again. The handling of the deferred test coverage has also been simplified.

This makes it possible to simplify the code even further in the future.

### CODE REVIEW

File: .teamcity/Gradle_Check/projects/FunctionalTestProject.kt [hunk 1]
```diff
@@ -7,30 +7,27 @@ import model.CIBuildModel
...
-            return@forEach
+            return@flatMap emptyList<FunctionalTest>()
+        }
+        val functionalTestsForCurrentBucket = bucket.forTestType(testConfig.testType).map {
+            buildTypeBucket -> FunctionalTest(model, testConfig, buildTypeBucket.getSubprojectNames(), stage, buildTypeBucket.name, buildTypeBucket.extraParameters())
         }
         if (bucket.shouldBeSkippedInStage(stage)) {
-            addMissingTestCoverage(testConfig)
-            return@forEach
+            deferredFunctionalTests.addAll(functionalTestsForCurrentBucket)
```
Comment (blindpirate) [inline .teamcity/Gradle_Check/projects/FunctionalTestProject.kt]: If I'm understanding correctly, this is modifying the passed parameter in `flatMap` (and caller's `map` [here](https://github.com/gradle/gradle/blob/859bd5b64aa7f3f613aac96b63d8094399221125/.teamcity/Gradle_Check/projects/StageProject.kt#L56)). Personally I don't like the side effect like this, can we somehow avoid it?

**Comment by wolfs** [comment]
> I.e. a parameter of the build changes from `-Dscan.tag.ReadyforNightly` to `-Dscan.tag.ReadyforMerge`. Can you check what causes this change?

Yes, see this commit message: https://github.com/gradle/gradle/pull/10728/commits/859bd5b64aa7f3f613aac96b63d8094399221125

I think I can fix it.

File: .teamcity/Gradle_Check/projects/FunctionalTestProject.kt (hunk 1 above)
Comment (wolfs) [reply .teamcity/Gradle_Check/projects/FunctionalTestProject.kt]: Now the `flatMap` doesn't have a side-effect any more.

File: .teamcity/Gradle_Check/projects/StageProject.kt [hunk 2]
```diff
@@ -5,83 +5,88 @@ import configurations.PerformanceTestCoordinator
...
+                if (!(stage.functionalTestsDependOnSpecificBuilds && stage.specificBuilds.contains(SpecificBuild.SanityCheck)) && stage.dependsOnSanityCheck) {
+                    functionalTestProject.addDependencyForAllBuildTypes(AbsoluteId(SanityCheck.buildTypeId(model)))
+                }
+                functionalTestProject
             }
-            if (!(stage.functionalTestsDependOnSpecificBuilds && stage.specificBuilds.contains(SpecificBuild.SanityCheck)) && stage.dependsOnSanityCheck) {
-                functionalTests.addDependencyForAllBuildTypes(AbsoluteId(SanityCheck.buildTypeId(model)))
-            }
+
+        functionalTestProjects.forEach {
```
[the remark this answers is not in the document]
Comment (wolfs) [reply .teamcity/Gradle_Check/projects/StageProject.kt]: We could, though I prefer not to have the side-effect within the `map`.

File: .teamcity/Gradle_Check/projects/StageProject.kt [hunk 3]
```diff
@@ -5,83 +5,88 @@ import configurations.PerformanceTestCoordinator
...
+            }
         }
-        if (!stage.performanceTests.isEmpty()) {
-            buildReportTab("Performance", "report-performance-performance-tests.zip!report/index.html")
+
+        specificBuildTypes = stage.specificBuilds.map {
+            it.create(model, stage)
         }
-    }
+        specificBuildTypes.forEach { buildType(it) }
```
[the remark this answers is not in the document]
Comment (wolfs) [reply .teamcity/Gradle_Check/projects/StageProject.kt]: Done.

File: .teamcity/Gradle_Check/projects/StageProject.kt [hunk 4]
```diff
@@ -5,83 +5,88 @@ import configurations.PerformanceTestCoordinator
...
 
-    val specificBuildTypes = stage.specificBuilds.map {
-        it.create(model, stage)
-    }
-    specificBuildTypes.forEach { buildType(it) }
+        performanceTests = stage.performanceTests.map { PerformanceTestCoordinator(model, it, stage) }
```
[the remark this answers is not in the document]
Comment (wolfs) [reply .teamcity/Gradle_Check/projects/StageProject.kt]: I think it is not worth it.

File: .teamcity/Gradle_Check/projects/StageProject.kt [hunk 5]
```diff
@@ -5,83 +5,88 @@ import configurations.PerformanceTestCoordinator
...
+
+    val functionalTests: List<FunctionalTest>
+
+    init {
+        features {
+            if (stage.specificBuilds.contains(SpecificBuild.SanityCheck)) {
+                buildReportTab("API Compatibility Report", "report-distributions-binary-compatibility-report.html")
+                buildReportTab("Incubating APIs Report", "incubation-reports/all-incubating.html")
+            }
+            if (!stage.performanceTests.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (wolfs) [reply .teamcity/Gradle_Check/projects/StageProject.kt]: Done.

**Comment by wolfs** [comment]
@blindpirate PTAL!

**Pull request merged**